lib/Parse/ParseDecl.cpp

```diff
@@ -3121,29 +3121,65 @@ ParserResult<PoundDiagnosticDecl> Parser::parseDeclPoundDiagnostic() {
   SourceLoc startLoc = 
     consumeToken(isError ? tok::pound_error : tok::pound_warning);
 
-  SourceLoc lParenLoc;
-  if (parseToken(tok::l_paren, lParenLoc, 
-                 diag::pound_diagnostic_expected, "(", isError))
-    return makeParserError();
+  SourceLoc lParenLoc = Tok.getLoc();
+  bool hadLParen = consumeIf(tok::l_paren);
 
   if (!Tok.is(tok::string_literal)) {
-    diagnose(startLoc, diag::pound_diagnostic_expected_string, isError);
+    // Catch #warning(oops, forgot the quotes)
+    SourceLoc wordsStartLoc = Tok.getLoc();
+
+    while (!Tok.isAtStartOfLine() && Tok.isNot(tok::r_paren)) {
```

Comment:
Don't forget to *consume* the `r_paren`!

Reply:
Ack! Thanks!